policy/reviewer/reviewer.go

```diff
@@ -177,42 +195,54 @@ func SelectReviewers(ctx context.Context, prctx pull.Context, result common.Resu
 
 		if child.ReviewRequestRule.Admins {
 			logger.Debug().Msg("Selecting from admins for review")
-			admins, err := selectAdmins(prctx)
+			adminTeams, err := selectAdmins(prctx)
 			if err != nil {
-				return nil, errors.Wrap(err, "failed to select admins")
+				return nil, nil, errors.Wrap(err, "failed to select admins")
 			}
-
-			for _, admin := range admins {
-				allUsers[admin] = struct{}{}
+			for team, members := range adminTeams {
+				allTeamsWithUsers[team] = members
+				for _, admin := range members {
+					allUsers[admin] = struct{}{}
+				}
 			}
 		}
 
-		var possibleReviewers []string
-		for u := range allUsers {
-			// Remove the author and any users who aren't collaborators
-			// since github will fail to assign _anyone_ if the request contains one of these
-			_, ok := collaboratorsToConsider[u]
-			if u != prctx.Author() && ok {
-				possibleReviewers = append(possibleReviewers, u)
+		allUsersKeys := make([]string, 0, len(allUsers))
+		for k := range allUsers {
+			allUsersKeys = append(allUsersKeys, k)
+		}
+
+		possibleReviewers := getPossibleReviewers(prctx, allUsersKeys, collaboratorsToConsider)
+		if child.GetMode() == common.RequestModePreferTeams {
+			var possibleTeams []string
+			teamMembersCount := 0
+			for team, users := range allTeamsWithUsers {
+				possibleReviewersInTeamCount := len(getPossibleReviewers(prctx, users, collaboratorsToConsider))
+				if possibleReviewersInTeamCount > 0 {
+					possibleTeams = append(possibleTeams, team)
+					teamMembersCount += possibleReviewersInTeamCount
+				}
+			}
+			if len(possibleTeams) > 0 && teamMembersCount >= child.ReviewRequestRule.RequiredCount {
+				logger.Debug().Msgf("Found %d total teams with a collective %d eligible members; requesting teams", len(possibleTeams), teamMembersCount)
+				teamsToRequest = append(teamsToRequest, possibleTeams...)
+				continue
 			}
 		}
 
 		if len(possibleReviewers) > 0 {
-			switch child.GetMode() {
-			case common.RequestModeAllUsers:
+			if child.GetMode() == common.RequestModeAllUsers {
 				logger.Debug().Msgf("Found %d total reviewers after removing author and non-collaborators; requesting all", len(possibleReviewers))
 				usersToRequest = append(usersToRequest, possibleReviewers...)
-			case common.RequestModeRandomUsers:
-				logger.Debug().Msgf("Found %d total candidates for review after removing author and non-collaborators; randomly selecting %d", len(possibleReviewers), child.ReviewRequestRule.RequiredCount)
-				randomSelection := selectRandomUsers(child.ReviewRequestRule.RequiredCount, possibleReviewers, r)
-				usersToRequest = append(usersToRequest, randomSelection...)
-			default:
-				return nil, fmt.Errorf("unknown mode '%s' supplied", child.ReviewRequestRule.Mode)
+				continue
```

Comment:
I know a lot of this structure already existed, but I'm finding this use of `continue` confusing given the size of the body of this loop. Do you think we could restructure the old `switch` statement so that each review mode still has a dedicated case?

I think this is easier if we remove the fallback behavior for teams, but maybe a `fallthrough` statement would make that clear if you want to keep the fallback behavior. Inverting the `possibleReviewers` length check and having that be the only `continue` could help as well.

Alternately, maybe there's a helper method to extract that implements the mode switching.